### PR TITLE
fix(core): clamp param_id buffer copy and cover helpers

### PR DESCRIPTION
**Pull request by Bartok9**
## Summary

- Clamp `param_id_to_message_buffer` copies to the MAVLink 16-byte field so over-long caller strings cannot write past the fixed array.
- Treat `nullptr` in `extract_safe_param_id` as empty (defensive).
- Add unit tests for null-terminated, full-16 non-terminated, clamp, and short round-trip paths.

## Motivation

`param_id_to_message_buffer` previously `memcpy`'d `param_id.length()` bytes into a 16-byte array. Callers that pass a longer string silently overflowed. Tests lock the clamp and the classic non-terminated 16-byte extract path.

## Verification

- Unit tests: `mavlink_parameter_helper_test.cpp` (registered in core `CMakeLists.txt`)
- Manual review of memcpy bounds
- Did NOT change: public ParamValue APIs or parameter wire encoding beyond buffer packing

---

**Agent-Owner:** `sera` · **Platform:** hermes · **Claim-TTL:** 24h

### CODE REVIEW

**Review by julianoes — COMMENTED** [review]
Please don't change the implementation like that. Add a commit that shows a valid unit test failure first and then fix it, and don't just move functions to the header.

File: cpp/src/mavsdk/core/mavlink_parameter_helper.hpp [hunk 1]
```diff
@@ -1,16 +1,40 @@
...
 #include <string>
 #include "mavlink_include.hpp"
 
 namespace mavsdk {
 
 static constexpr size_t PARAM_ID_LEN = 16;
 
-[[nodiscard]] std::string extract_safe_param_id(const char* param_id);
+// Inline so unit tests (and clients) always see definitions without depending
+// on shared-library export edge cases for small buffer helpers.
```
Comment (julianoes) [inline cpp/src/mavsdk/core/mavlink_parameter_helper.hpp]: Nope, disagree with this, no good reason to change it.

**Review by julianoes — CHANGES_REQUESTED** [review]

**Comment by Bartok9** [comment]
Thanks for the review @julianoes — both points addressed:

1. **Failing test first, then fix.** Split into two commits: [`test(core)`](../commits) adds the unit tests against the *current* implementation (the clamp and nullptr cases fail — the `memcpy` writes past the 16-byte array on long input, and `extract_safe_param_id` dereferences null), and the follow-up `fix(core)` commit makes them pass.
2. **No functions moved to the header.** Implementations stay in `mavlink_parameter_helper.cpp`; the header keeps declarations only, unchanged from main. The fix is just the length clamp (`std::min(param_id.length(), PARAM_ID_LEN)`) plus a null guard.

Diff is now just the two `.cpp` helper changes + the test file + one CMake test-registration line. Let me know if you'd like the null-guard dropped if you consider it out of scope.

**Comment by julianoes** [comment]
Fix the conflicts first please.

**Comment by Bartok9** [comment]
Rebased on latest `main` to clear the conflict (only a one-line `CMakeLists.txt` test-registration overlap with the new `param_value_xml_test.cpp` entry). Structure unchanged: failing `test(core)` commit first, then the `fix(core)` clamp + null-guard. Branch is green/mergeable again — ready when you are.

**Review by julianoes — DISMISSED** [review]
Ok, now I can review it and actually agree with the change!

File: cpp/src/mavsdk/core/mavlink_parameter_helper.hpp (hunk 1 above)
Comment (Bartok9) [reply cpp/src/mavsdk/core/mavlink_parameter_helper.hpp]: Agreed — that header move is gone. Implementations stay in `mavlink_parameter_helper.cpp`; the `.hpp` is untouched vs `main`. Thanks for the clear rejection.

**Comment by Bartok9** [comment]
Thanks @julianoes — glad the red-green split + keeping the implementations in the `.cpp` worked for you. Appreciate the clear guidance (and the pushback on the header move); we’ll run that pattern on future MAVSDK fixes.

**Comment by julianoes** [comment]
Fuzzing doesn't link. I wonder why.

**Comment by Bartok9** [comment]
@julianoes good catch on the fuzzing link failure — same hidden-visibility pitfall as the other unit-test helpers.

Root cause: `extract_safe_param_id` / `param_id_to_message_buffer` lived in `libmavsdk` with default hidden visibility, so `unit_tests_runner` (and the fuzz build) got undefined references at link time.

Fix just landed: mark both free functions `MAVSDK_TEST_EXPORT` in `mavlink_parameter_helper.hpp` (same pattern as `string_utils` / `fs_utils` / the flight_mode converters). Implementations stay in the `.cpp`; no header body move. CI on the latest head should go green once the matrix reruns.

**Comment by Bartok9** [comment]
@julianoes good catch on the fuzzing link failure — same hidden-visibility pitfall as the other unit-test helpers.

Root cause: `extract_safe_param_id` / `param_id_to_message_buffer` lived in `libmavsdk` with default hidden visibility, so `unit_tests_runner` (and the fuzz build) got undefined references at link time.

Fix just landed: mark both free functions `MAVSDK_TEST_EXPORT` in `mavlink_parameter_helper.hpp` (same pattern as `string_utils` / `fs_utils` / the flight_mode converters). Implementations stay in the `.cpp`; no header body move. CI on the latest head should go green once the matrix reruns.

**Comment by Bartok9** [comment]
Consolidating per maintainer feedback into a single core tests PR: https://github.com/mavlink/MAVSDK/pull/3004

Closing this one to cut churn/conflicts. Thanks!
